File: swift-bridge/Sources/CoreAnimationBridge/CATiledLayer.swift
import Foundation
import QuartzCore

@_cdecl("ca_tiled_layer_new")
public func ca_tiled_layer_new() -> UnsafeMutableRawPointer? {
    caRetain(CATiledLayer())
}

@_cdecl("ca_tiled_layer_get_levels_of_detail")
public func ca_tiled_layer_get_levels_of_detail(_ handle: UnsafeMutableRawPointer?) -> Int {
    guard let layer: CATiledLayer = caBorrow(handle) else { return 0 }
    return Int(layer.levelsOfDetail)
}

@_cdecl("ca_tiled_layer_set_levels_of_detail")
public func ca_tiled_layer_set_levels_of_detail(_ handle: UnsafeMutableRawPointer?, _ value: Int) {
    guard let layer: CATiledLayer = caBorrow(handle) else { return }
    layer.levelsOfDetail = max(1, value)
}

@_cdecl("ca_tiled_layer_get_levels_of_detail_bias")
public func ca_tiled_layer_get_levels_of_detail_bias(_ handle: UnsafeMutableRawPointer?) -> Int {
    guard let layer: CATiledLayer = caBorrow(handle) else { return 0 }
    return Int(layer.levelsOfDetailBias)
}

@_cdecl("ca_tiled_layer_set_levels_of_detail_bias")
public func ca_tiled_layer_set_levels_of_detail_bias(_ handle: UnsafeMutableRawPointer?, _ value: Int) {
    guard let layer: CATiledLayer = caBorrow(handle) else { return }
    layer.levelsOfDetailBias = max(0, value)
}

@_cdecl("ca_tiled_layer_get_tile_size")
public func ca_tiled_layer_get_tile_size(_ handle: UnsafeMutableRawPointer?, _ outSize: UnsafeMutableRawPointer?) -> Bool {
    guard let layer: CATiledLayer = caBorrow(handle) else { return false }
    return caWriteSize(layer.tileSize, out: outSize)
}

@_cdecl("ca_tiled_layer_set_tile_size")
public func ca_tiled_layer_set_tile_size(_ handle: UnsafeMutableRawPointer?, _ width: Double, _ height: Double) {
    guard let layer: CATiledLayer = caBorrow(handle) else { return }
    layer.tileSize = CGSize(width: width, height: height)
}

File: src/ca_tiled_layer.rs
use std::ops::Deref;

use apple_cf::cg::CGSize;

use crate::layer::{Layer, LayerLike};

#[derive(Debug, Clone)]
pub struct TiledLayer {
    inner: Layer,
}

impl TiledLayer {
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Layer::from_raw(crate::ffi::ca_tiled_layer_new()) }.map(|inner| Self { inner })
    }

    #[must_use]
    pub fn levels_of_detail(&self) -> usize {
        unsafe { crate::ffi::ca_tiled_layer_get_levels_of_detail(self.as_layer_ptr()) }
    }

    pub fn set_levels_of_detail(&self, value: usize) {
        unsafe { crate::ffi::ca_tiled_layer_set_levels_of_detail(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn levels_of_detail_bias(&self) -> usize {
        unsafe { crate::ffi::ca_tiled_layer_get_levels_of_detail_bias(self.as_layer_ptr()) }
    }

    pub fn set_levels_of_detail_bias(&self, value: usize) {
        unsafe { crate::ffi::ca_tiled_layer_set_levels_of_detail_bias(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn tile_size(&self) -> CGSize {
        let mut size = CGSize::zero();
        let ok = unsafe {
            crate::ffi::ca_tiled_layer_get_tile_size(
                self.as_layer_ptr(),
                (&mut size as *mut CGSize).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            size
        } else {
            CGSize::zero()
        }
    }

    pub fn set_tile_size(&self, size: CGSize) {
        unsafe {
            crate::ffi::ca_tiled_layer_set_tile_size(self.as_layer_ptr(), size.width, size.height)
        };
    }
}

impl Deref for TiledLayer {
    type Target = Layer;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl LayerLike for TiledLayer {
    fn as_layer_ptr(&self) -> *mut core::ffi::c_void {
        self.inner.as_ptr()
    }
}
